File: packages/kimdb-client/src/websocket.ts
/**
 * KimDB WebSocket Client - Real-time synchronization
 * @module @kimdb/client/websocket
 */

import { EventEmitter } from 'events';
import WebSocket from 'ws';

export interface SubscribeMessage {
  type: 'subscribe';
  collection: string;
}

export interface DocSubscribeMessage {
  type: 'doc.subscribe';
  collection: string;
  docId: string;
}

export interface DocUpdateMessage {
  type: 'doc.update';
  collection: string;
  docId: string;
  data: Record<string, unknown>;
  nodeId: string;
}

export interface DocUndoMessage {
  type: 'doc.undo';
  collection: string;
  docId: string;
  nodeId: string;
}

export interface DocRedoMessage {
  type: 'doc.redo';
  collection: string;
  docId: string;
  nodeId: string;
}

export interface PresenceUpdateMessage {
  type: 'presence.update';
  collection: string;
  docId: string;
  nodeId: string;
  presence: {
    cursor?: { line: number; column: number };
    selection?: { start: number; end: number };
    name?: string;
  };
}

export interface PingMessage {
  type: 'ping';
}

export type WebSocketMessage =
  | SubscribeMessage
  | DocSubscribeMessage
  | DocUpdateMessage
  | DocUndoMessage
  | DocRedoMessage
  | PresenceUpdateMessage
  | PingMessage;

export interface DocSyncedMessage {
  type: 'doc.synced';
  collection: string;
  docId: string;
  data: Record<string, unknown>;
  _version: number;
}

export interface PresenceChangedMessage {
  type: 'presence.changed';
  docId: string;
  nodeId: string;
  presence: Record<string, unknown>;
}

export interface PongMessage {
  type: 'pong';
  timestamp: number;
}

/**
 * KimDB WebSocket Client for real-time synchronization
 */
export class KimDBWebSocket extends EventEmitter {
  private ws: WebSocket | null = null;
  private url: string;
  private nodeId: string;
  private reconnectInterval: number;
  private heartbeatInterval: NodeJS.Timeout | null = null;
  private isConnected: boolean = false;

  constructor(url: string, nodeId?: string) {
    super();
    this.url = url;
    this.nodeId = nodeId || `client-${Math.random().toString(36).substr(2, 9)}`;
    this.reconnectInterval = 5000;
  }

  /**
   * Connect to WebSocket server
   */
  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.url);

        this.ws.onopen = () => {
          this.isConnected = true;
          console.log(`[KimDB] WebSocket connected (nodeId: ${this.nodeId})`);
          this.startHeartbeat();
          this.emit('connected');
          resolve();
        };

        this.ws.onmessage = (event) => {
          try {
            const message = JSON.parse(event.data as string);
            this.handleMessage(message);
          } catch (error) {
            console.error('[KimDB] Failed to parse message:', error);
          }
        };

        this.ws.onerror = (error) => {
          console.error('[KimDB] WebSocket error:', error);
          this.emit('error', error);
          reject(error);
        };

        this.ws.onclose = () => {
          this.isConnected = false;
          console.log('[KimDB] WebSocket disconnected');
          this.stopHeartbeat();
          this.emit('disconnected');
          this.reconnect();
        };
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Handle incoming WebSocket messages
   */
  private handleMessage(message: any): void {
    switch (message.type) {
      case 'subscribed':
        this.emit('subscribed', message.collection);
        break;

      case 'doc.synced':
        this.emit('doc.synced', {
          collection: message.collection,
          docId: message.docId,
          data: message.data,
          version: message._version,
        });
        break;

      case 'doc.updated':
        this.emit('doc.updated', {
          docId: message.docId,
          success: message.success,
          version: message._version,
        });
        break;

      case 'presence.changed':
        this.emit('presence.changed', {
          docId: message.docId,
          nodeId: message.nodeId,
          presence: message.presence,
        });
        break;

      case 'pong':
        this.emit('pong', message.timestamp);
        break;

      case 'error':
        console.error('[KimDB] Server error:', message.error);
        this.emit('error', new Error(message.error));
        break;

      default:
        console.warn('[KimDB] Unknown message type:', message.type);
    }
  }

  /**
   * Subscribe to collection updates
   */
  subscribe(collection: string): void {
    if (!this.isConnected) {
      throw new Error('WebSocket not connected');
    }

    const message: SubscribeMessage = { type: 'subscribe', collection };
    this.ws?.send(JSON.stringify(message));
  }

  /**
   * Subscribe to specific document
   */
  subscribeDocument(collection: string, docId: string): void {
    if (!this.isConnected) {
      throw new Error('WebSocket not connected');
    }

    const message: DocSubscribeMessage = {
      type: 'doc.subscribe',
      collection,
      docId,
    };
    this.ws?.send(JSON.stringify(message));
  }

  /**
   * Update document with CRDT sync
   */
  updateDocument(
    collection: string,
    docId: string,
    data: Record<string, unknown>
  ): void {
    if (!this.isConnected) {
      throw new Error('WebSocket not connected');
    }

    const message: DocUpdateMessage = {
      type: 'doc.update',
      collection,
      docId,
      data,
      nodeId: this.nodeId,
    };
    this.ws?.send(JSON.stringify(message));
  }

  /**
   * Undo last operation
   */
  undo(collection: string, docId: string): void {
    if (!this.isConnected) {
      throw new Error('WebSocket not connected');
    }

    const message: DocUndoMessage = {
      type: 'doc.undo',
      collection,
      docId,
      nodeId: this.nodeId,
    };
    this.ws?.send(JSON.stringify(message));
  }

  /**
   * Redo operation
   */
  redo(collection: string, docId: string): void {
    if (!this.isConnected) {
      throw new Error('WebSocket not connected');
    }

    const message: DocRedoMessage = {
      type: 'doc.redo',
      collection,
      docId,
      nodeId: this.nodeId,
    };
    this.ws?.send(JSON.stringify(message));
  }

  /**
   * Update presence information
   */
  updatePresence(
    collection: string,
    docId: string,
    presence: Record<string, unknown>
  ): void {
    if (!this.isConnected) {
      throw new Error('WebSocket not connected');
    }

    const message: PresenceUpdateMessage = {
      type: 'presence.update',
      collection,
      docId,
      nodeId: this.nodeId,
      presence: presence as any,
    };
    this.ws?.send(JSON.stringify(message));
  }

  /**
   * Send heartbeat ping
   */
  private ping(): void {
    if (this.isConnected) {
      const message: PingMessage = { type: 'ping' };
      this.ws?.send(JSON.stringify(message));
    }
  }

  /**
   * Start heartbeat
   */
  private startHeartbeat(): void {
    this.heartbeatInterval = setInterval(() => this.ping(), 30000);
  }

  /**
   * Stop heartbeat
   */
  private stopHeartbeat(): void {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
    }
  }

  /**
   * Attempt reconnection
   */
  private reconnect(): void {
    console.log('[KimDB] Reconnecting in 5 seconds...');
    setTimeout(() => {
      this.connect().catch((error) => {
        console.error('[KimDB] Reconnection failed:', error);
      });
    }, this.reconnectInterval);
  }

  /**
   * Disconnect WebSocket
   */
  disconnect(): void {
    this.stopHeartbeat();
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.isConnected = false;
  }

  /**
   * Check connection status
   */
  connected(): boolean {
    return this.isConnected;
  }
}

export default KimDBWebSocket;

File: packages/kimdb-go/websocket.go
package kimdb

import (
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// WebSocketMessage represents a WebSocket message
type WebSocketMessage struct {
	Type       string                 `json:"type"`
	Collection string                 `json:"collection,omitempty"`
	DocID      string                 `json:"docId,omitempty"`
	Data       map[string]interface{} `json:"data,omitempty"`
	NodeID     string                 `json:"nodeId,omitempty"`
	Presence   map[string]interface{} `json:"presence,omitempty"`
	Timestamp  int64                  `json:"timestamp,omitempty"`
}

// WebSocketEvent represents an event
type WebSocketEvent struct {
	Type   string
	Data   map[string]interface{}
	Error  error
}

// WebSocket client for real-time synchronization
type WebSocket struct {
	url               string
	nodeID            string
	conn              *websocket.Conn
	connected         bool
	mu                sync.RWMutex
	messageHandlers   map[string][]func(WebSocketEvent)
	heartbeatTicker   *time.Ticker
	reconnectInterval time.Duration
}

// NewWebSocket creates a new WebSocket client
func NewWebSocket(url string, nodeID *string) *WebSocket {
	id := *nodeID
	if id == "" {
		id = fmt.Sprintf("client-%d", rand.Int63())
	}

	return &WebSocket{
		url:               url,
		nodeID:            id,
		messageHandlers:   make(map[string][]func(WebSocketEvent)),
		reconnectInterval: 5 * time.Second,
	}
}

// On registers an event handler
func (ws *WebSocket) On(event string, handler func(WebSocketEvent)) {
	ws.mu.Lock()
	defer ws.mu.Unlock()

	ws.messageHandlers[event] = append(ws.messageHandlers[event], handler)
}

// Connect connects to the WebSocket server
func (ws *WebSocket) Connect() error {
	dialer := websocket.Dialer{
		HandshakeTimeout: 10 * time.Second,
	}

	conn, _, err := dialer.Dial(ws.url, nil)
	if err != nil {
		return fmt.Errorf("failed to connect: %w", err)
	}

	ws.mu.Lock()
	ws.conn = conn
	ws.connected = true
	ws.mu.Unlock()

	log.Printf("[KimDB] WebSocket connected (nodeId: %s)", ws.nodeID)
	ws.emit(WebSocketEvent{Type: "connected"})

	ws.startHeartbeat()
	go ws.readMessages()

	return nil
}

// readMessages reads incoming messages
func (ws *WebSocket) readMessages() {
	defer ws.disconnect()

	for {
		ws.mu.RLock()
		if !ws.connected {
			ws.mu.RUnlock()
			return
		}
		conn := ws.conn
		ws.mu.RUnlock()

		var msg WebSocketMessage
		err := conn.ReadJSON(&msg)
		if err != nil {
			ws.mu.Lock()
			ws.connected = false
			ws.mu.Unlock()
			log.Printf("[KimDB] WebSocket error: %v", err)
			ws.emit(WebSocketEvent{Type: "disconnected"})
			return
		}

		ws.handleMessage(msg)
	}
}

// handleMessage processes an incoming message
func (ws *WebSocket) handleMessage(msg WebSocketMessage) {
	event := WebSocketEvent{
		Type: msg.Type,
		Data: make(map[string]interface{}),
	}

	switch msg.Type {
	case "subscribed":
		event.Data["collection"] = msg.Collection
	case "doc.synced":
		event.Data["collection"] = msg.Collection
		event.Data["docId"] = msg.DocID
		event.Data["data"] = msg.Data
		event.Data["version"] = msg.Timestamp
	case "doc.updated":
		event.Data["docId"] = msg.DocID
		event.Data["success"] = true
	case "presence.changed":
		event.Data["docId"] = msg.DocID
		event.Data["nodeId"] = msg.NodeID
		event.Data["presence"] = msg.Presence
	case "pong":
		event.Data["timestamp"] = msg.Timestamp
	case "error":
		event.Error = fmt.Errorf("%v", msg.Data["error"])
	}

	ws.emit(event)
}

// emit emits an event to all registered handlers
func (ws *WebSocket) emit(event WebSocketEvent) {
	ws.mu.RLock()
	handlers := ws.messageHandlers[event.Type]
	ws.mu.RUnlock()

	for _, handler := range handlers {
		go handler(event)
	}
}

// Subscribe subscribes to collection updates
func (ws *WebSocket) Subscribe(collection string) error {
	ws.mu.RLock()
	if !ws.connected {
		ws.mu.RUnlock()
		return fmt.Errorf("not connected")
	}
	conn := ws.conn
	ws.mu.RUnlock()

	msg := WebSocketMessage{
		Type:       "subscribe",
		Collection: collection,
	}

	return conn.WriteJSON(msg)
}

// SubscribeDocument subscribes to a specific document
func (ws *WebSocket) SubscribeDocument(collection, docID string) error {
	ws.mu.RLock()
	if !ws.connected {
		ws.mu.RUnlock()
		return fmt.Errorf("not connected")
	}
	conn := ws.conn
	ws.mu.RUnlock()

	msg := WebSocketMessage{
		Type:       "doc.subscribe",
		Collection: collection,
		DocID:      docID,
	}

	return conn.WriteJSON(msg)
}

// UpdateDocument updates a document
func (ws *WebSocket) UpdateDocument(collection, docID string, data map[string]interface{}) error {
	ws.mu.RLock()
	if !ws.connected {
		ws.mu.RUnlock()
		return fmt.Errorf("not connected")
	}
	conn := ws.conn
	ws.mu.RUnlock()

	msg := WebSocketMessage{
		Type:       "doc.update",
		Collection: collection,
		DocID:      docID,
		Data:       data,
		NodeID:     ws.nodeID,
	}

	return conn.WriteJSON(msg)
}

// Undo undoes the last operation
func (ws *WebSocket) Undo(collection, docID string) error {
	ws.mu.RLock()
	if !ws.connected {
		ws.mu.RUnlock()
		return fmt.Errorf("not connected")
	}
	conn := ws.conn
	ws.mu.RUnlock()

	msg := WebSocketMessage{
		Type:       "doc.undo",
		Collection: collection,
		DocID:      docID,
		NodeID:     ws.nodeID,
	}

	return conn.WriteJSON(msg)
}

// Redo redoes an operation
func (ws *WebSocket) Redo(collection, docID string) error {
	ws.mu.RLock()
	if !ws.connected {
		ws.mu.RUnlock()
		return fmt.Errorf("not connected")
	}
	conn := ws.conn
	ws.mu.RUnlock()

	msg := WebSocketMessage{
		Type:       "doc.redo",
		Collection: collection,
		DocID:      docID,
		NodeID:     ws.nodeID,
	}

	return conn.WriteJSON(msg)
}

// UpdatePresence updates presence information
func (ws *WebSocket) UpdatePresence(collection, docID string, presence map[string]interface{}) error {
	ws.mu.RLock()
	if !ws.connected {
		ws.mu.RUnlock()
		return fmt.Errorf("not connected")
	}
	conn := ws.conn
	ws.mu.RUnlock()

	msg := WebSocketMessage{
		Type:       "presence.update",
		Collection: collection,
		DocID:      docID,
		NodeID:     ws.nodeID,
		Presence:   presence,
	}

	return conn.WriteJSON(msg)
}

// startHeartbeat starts the heartbeat
func (ws *WebSocket) startHeartbeat() {
	ws.heartbeatTicker = time.NewTicker(30 * time.Second)

	go func() {
		for range ws.heartbeatTicker.C {
			ws.ping()
		}
	}()
}

// ping sends a ping message
func (ws *WebSocket) ping() {
	ws.mu.RLock()
	if !ws.connected {
		ws.mu.RUnlock()
		return
	}
	conn := ws.conn
	ws.mu.RUnlock()

	msg := WebSocketMessage{Type: "ping"}
	if err := conn.WriteJSON(msg); err != nil {
		log.Printf("[KimDB] Ping failed: %v", err)
	}
}

// disconnect disconnects from the server
func (ws *WebSocket) disconnect() {
	ws.mu.Lock()
	defer ws.mu.Unlock()

	ws.connected = false

	if ws.heartbeatTicker != nil {
		ws.heartbeatTicker.Stop()
	}

	if ws.conn != nil {
		ws.conn.Close()
	}

	log.Println("[KimDB] WebSocket disconnected")
}

// Disconnect disconnects from the server
func (ws *WebSocket) Disconnect() {
	ws.disconnect()
}

// IsConnected checks if connected
func (ws *WebSocket) IsConnected() bool {
	ws.mu.RLock()
	defer ws.mu.RUnlock()
	return ws.connected
}
